### ISSUE TITLE
Deadly Bug found and solved !!! Must be fixed !!!

### ISSUE THREAD

**Issue by TZYSJTU** [zoubohao/DenoisingDiffusionProbabilityModel-ddpm-#14]
The beta calculation is wrong!
Compared with the official DDPM code,
three important lines are missing in this unofficiial version !

Below are the official codes of beta's calculation:
    def linear_beta_schedule(timesteps):
        """
        linear schedule, proposed in original ddpm paper
        """
        scale = 1000 / timesteps
        beta_start = scale * 0.0001
        beta_end = scale * 0.02
        return torch.linspace(beta_start, beta_end, timesteps, dtype = torch.float64)

the scale related three lines are missing in your code,
which  leads to a deadly bug：

when using T = 100 or 50 or 20, much less than 1000
the calcuated x_T is not standard Gaussian!
Therefore, you can't sample from a standard Gaussian after training.

You can easily check my word by calculating the mean and var of x_T !

**Comment by jinzi98** [comment]
I think T need to be a big number to make x_T approximately equal to a standard Gaussian.  20, 50 and 100 are too small so x_T is not Gaussian.

**Comment by TZYSJTU** [comment]
wrong! What ever T=50, 100, 200, 1000....
x_T is standard Gaussian！
&nbsp;


Einsames&nbsp;Brandenburg
***@***.***



&nbsp;




------------------&nbsp;原始邮件&nbsp;------------------
发件人: ***@***.***&gt;; 
发送时间: 2023年3月1日(星期三) 下午5:45
收件人: ***@***.***&gt;; 
抄送: "Einsames ***@***.***&gt;; ***@***.***&gt;; 
主题: Re: [zoubohao/DenoisingDiffusionProbabilityModel-ddpm-] Deadly Bug found and solved !!! Must be fixed !!! (Issue #14)





 
I think T need to be a big number to make x_T approximately equal to a standard Gaussian.  20, 50 and 100 are too small so x_T is not Gaussian.

**Comment by HIT-lcccc** [comment]
Here is not deadly bug. In official code, `scale` is related to `timesteps` but if read the code carefully, you wiil find that `timesteps` is set to 1000 in `modelConfig`. I also don't think X_T is standard Gaussian when using T = 100 or 50 or 20 much less than 1000. In other words, if it becomes standard Gaussian in preceding step, there is no need to set T as 1000 or more.

**Comment by TZYSJTU** [comment]
You are totally wrong. I think you should so some work about math.&nbsp;
&nbsp;


Einsames&nbsp;Brandenburg
***@***.***



&nbsp;




------------------&nbsp;原始邮件&nbsp;------------------
发件人: ***@***.***&gt;; 
发送时间: 2023年3月23日(星期四) 下午3:45
收件人: ***@***.***&gt;; 
抄送: "Einsames ***@***.***&gt;; ***@***.***&gt;; 
主题: Re: [zoubohao/DenoisingDiffusionProbabilityModel-ddpm-] Deadly Bug found and solved !!! Must be fixed !!! (Issue #14)





 
Here is not deadly bug. In official code, scale is related to timesteps but if read the code carefully, you wiil find that timesteps is set to 1000 in modelConfig. I also don't think X_T is standard Gaussian when using T = 100 or 50 or 20 much less than 1000. In other words, if it becomes standard Gaussian in preceding step, there is no need to set T as 1000 or more.

**Comment by ljh430** [comment]
> You are totally wrong. I think you should so some work about math.&nbsp; &nbsp; Einsames&nbsp;Brandenburg ***@***.*** &nbsp;
> […](#)
> ------------------&nbsp;原始邮件&nbsp;------------------ 发件人: ***@***.***&gt;; 发送时间: 2023年3月23日(星期四) 下午3:45 收件人: ***@***.***&gt;; 抄送: "Einsames ***@***.***&gt;; ***@***.***&gt;; 主题: Re: [zoubohao/DenoisingDiffusionProbabilityModel-ddpm-] Deadly Bug found and solved !!! Must be fixed !!! (Issue #14) Here is not deadly bug. In official code, scale is related to timesteps but if read the code carefully, you wiil find that timesteps is set to 1000 in modelConfig. I also don't think X_T is standard Gaussian when using T = 100 or 50 or 20 much less than 1000. In other words, if it becomes standard Gaussian in preceding step, there is no need to set T as 1000 or more. — Reply to this email directly, view it on GitHub, or unsubscribe. You are receiving this because you authored the thread.Message ID: ***@***.***&gt;

so you believe  we need to add below code:
scale = 1000 / timesteps
beta_start = scale * 0.0001
beta_end = scale * 0.02

to scale the beta plan when set T<1000, right?

**Comment by cainiao123s** [comment]
I've already looked at the beta section of this code base, and when T is set to 1000, the beta is the same as the source code, and when T is not 1000, the beta is linearly amplified, which seems to me to add enough noise to X_0 within a finite T：class GaussianDiffusionTrainer(nn.Module):
    def __init__(self, model, beta_1, beta_T, T):
        super().__init__()

        self.model = model
        self.T = T

        self.register_buffer(
            'betas', torch.linspace(beta_1, beta_T, T).double())
        alphas = 1. - self.betas
        alphas_bar = torch.cumprod(alphas, dim=0)

        # calculations for diffusion q(x_t | x_{t-1}) and others
        self.register_buffer(
            'sqrt_alphas_bar', torch.sqrt(alphas_bar))
        self.register_buffer(
            'sqrt_one_minus_alphas_bar', torch.sqrt(1. - alphas_bar))
